mlflow/projects/__init__.py

```diff
@@ -279,58 +300,102 @@ def _maybe_create_conda_env(conda_env_path):
+        mlflow_run_arr.extend(["--storage-dir", storage_dir])
+    if not use_conda:
+        mlflow_run_arr.append("--no-conda")
+    if run_id:
+        mlflow_run_arr.extend(["--run-id", run_id])
+    if parameters:
+        for key, value in parameters.items():
+            mlflow_run_arr.extend(["-P", "%s=%s" % (key, value)])
+    return mlflow_run_arr
+
+
+def _run_mlflow_run_cmd(mlflow_run_arr, env_map):
+    """
+    Invokes `mlflow run` in a subprocess using a special entry-point mode, which will in turn run
+    the entry point in a child process. Returns a handle to the subprocess.Popen launched to invoke
+    `mlflow run`.
+    """
+    final_env = os.environ.copy()
+    final_env.update(env_map)
+    # TODO: maybe write the output of asynchronous local runs e.g. as artifacts for
+    # debugging, we currently just drop it
+    # Launch `mlflow run` command as the leader of its own process group so that we can do a
+    # best-effort cleanup of all its descendant processes if needed
+    return subprocess.Popen(
+        mlflow_run_arr, env=final_env, universal_newlines=True,
+        stderr=open(os.devnull, "w"), stdout=open(os.devnull, "w"), preexec_fn=os.setsid)
+
+
+def _create_run(uri, experiment_id, work_dir, entry_point, parameters):
     active_run = tracking._create_run(
-        experiment_id=experiment_id, source_name=project.uri,
+        experiment_id=experiment_id, source_name=_expand_uri(uri),
         source_version=tracking._get_git_commit(work_dir), entry_point_name=entry_point,
         source_type=SourceType.PROJECT)
     if parameters is not None:
         for key, value in parameters.items():
             active_run.log_param(Param(key, value))
+    return active_run
+
+
+def _run_local(
+        work_dir, entry_point, parameters, experiment_id, use_conda, storage_dir, run_id):
+    """
+    Run an MLflow project from the given URI in a new directory.
+
+    Supports downloading projects from Git URIs with a specified version, or copying them from
+    the file system. For Git-based projects, a commit can be specified as the `version`.
+    """
+    eprint("=== Asynchronously launching MLflow run with ID %s ===" % run_id)
     # Add the run id into a magic environment variable that the subprocess will read,
     # causing it to reuse the run.
     env_map = {
-        tracking._RUN_ID_ENV_VAR: active_run.run_info.run_uuid,
+        tracking._RUN_ID_ENV_VAR: run_id,
         tracking._TRACKING_URI_ENV_VAR: tracking.get_tracking_uri(),
         tracking._EXPERIMENT_ID_ENV_VAR: str(experiment_id),
     }
-
-    commands.append(run_project_command)
-    command = " && ".join(commands)
-    eprint("=== Running command '%s' in run with ID '%s' === "
-           % (command, active_run.run_info.run_uuid))
-
-    return _launch_local_run(
-        active_run, command, work_dir, env_map, stream_output=block)
+    # Invoke `mlflow run` with a special mode, which will in turn run the entry point command in
```

Comment:
Can remove the special mode thing, and just mention that this will always run in blocking mode.